tests/baselines/reference/mappedTypes4.errors.txt

```diff
@@ -0,0 +1,77 @@
+tests/cases/conformance/types/mapped/mappedTypes4.ts(11,23): error TS2407: The right-hand side of a 'for...in' statement must be of type 'any', an object type or a type parameter, but here has type '(T & null) | (T & object)'.
+
+
+==== tests/cases/conformance/types/mapped/mappedTypes4.ts (1 errors) ====
+    type Box<T> = {
+    };
+    
+    type Boxified<T> = {
+        [P in keyof T]: Box<T[P]>;
+    };
+    
+    function boxify<T>(obj: T): Boxified<T> {
+        if (typeof obj === "object") {
+            let result = {} as Boxified<T>;
+            for (let k in obj) {
+                          ~~~
+!!! error TS2407: The right-hand side of a 'for...in' statement must be of type 'any', an object type or a type parameter, but here has type '(T & null) | (T & object)'.

```

Comment:
[This line of change](https://github.com/microsoft/TypeScript/pull/49091/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R21386) can fix this new error (we get the fact-based-nonnullable type of the thing on the right of `in`, but `T & null` isn't recognized as a null type without something like this). IMO, it's a pretty straightforward thing to do - deriving falsy flags from apparent intersection constituents.

Reply:
Our check of the right hand expression in `for-in` is this wonky logic in `checkForInStatement`:

```ts
    if (rightType === neverType || !isTypeAssignableToKind(rightType, TypeFlags.NonPrimitive | TypeFlags.InstantiableNonPrimitive)) {
```

That's what needs to be fixed. The error we report now actually seems fine (i.e. we probably don't want to permit `for-in` on something that can be `null`), the problem is really that there wasn't an error before. Alternatively you could argue _any_ type should be valid given that `for-in` never throws an error.

Reply:
Sure, but `rightType` is defined using `getNonNullableTypeIfNeeded`, meaning null and undefined always have been and should be allowed - that a type with a `null` comes out defies expectation, interesting rules on what types are allowed beyond that notwithstanding. 😛

Reply:
This is still a problem, we _allow_ `null` here, it's just filtered by how the type is constructed and not the assignability check, this shouldn't be an error.

Reply:
The real problem is our strange `for-in` checking. That's what #49193 is for.

Reply:
No, I don't think it is. Removing the error doesn't fix the problem - this code will still be broken by this change even if the error is removed:
```ts
// function which remaps `null` to a nonnull singleton, overloaded so already non-null types don't need to consider it
declare function process<T>(arg: NonNullable<T>): NonNullable<T>; // no `null` marker in output if no `null` in input
declare function process<T>(arg?: T): NonNullable<T> | {nullyMarker: true};

declare function throwOnNull<T>(arg: T): asserts arg is NonNullable<T>;

function f<T>(x?: T) {
    throwOnNull(x);
    if (typeof x === "object") {
        for (const k in x!) {
            const res = process(x); // chooses nullyMarker overload by accident with this PR, previously correctly selected nonnullable overload
        }
    }
}
```
Today, `res` is `NonNullable<T>`, however with this change to `NonNullable` and `typeof` object narrowing, `res` incorrectly becomes `NonNullable<T> | {nullyMarker: true}` (because the `NonNullable` now fails to remove the null branch of the type because of the incorrect type fact calculation so we think `x` can still be `null` in the loop body when it in fact cannot). That's the implication of [this line of the type baseline diff](https://github.com/microsoft/TypeScript/pull/49119/files#diff-38044aa33e96fed3f3a0d00780120762962e188b30c78977a4febf01e1d3b9beR33), which removing the error won't change.

Reply:
I get that, and I think that's an issue for another PR. I mean, this is already broken today:

```ts
function f<T>(x: T & null | T & object) {
    let res = x || {a: 1}
    "a" in res; // This PR issues an error - incorrectly thinks `res` can still be `null`, despite definitely being assigned
}
```

I don't want to solve unrelated problems in this PR. Just like I'm not fixing the `T` assignable to `{}` in this PR.

Reply:
Why are we even doing `getNonNullableTypeIfNeeded` in `for-in` when the statement is perfectly happy with a `null` or `undefined` right hand side value? And, even worse, why are we removing `null` and `undefined` from the right hand side in CFA when the statement doesn't actually do that. Those are issues that need fixing. In another PR!

Reply:
> I don't want to solve unrelated problems in this PR. 

I mean, the only thing that only *needed* to be done to make `typeof x === "object"` narrow generics was change the code to do just that - the bulk of this PR is followup (or perhaps prerequisite) stuff to make that a less breaky change, which I would have assumed includes making these other cases _also_ not regress, hence my comments. As-is, I've presented a few real and hopefully reasonable code examples which function fine today, but with only these changes, will be broken, each an immediate regression we'll have to file bugs for and fix, as is the test change highlighted here, imo. Certainly, we had control flow narrowing issues with `unknown` before this PR, too, but there're in this PR because fixing them was relevant for the typeof object narrowing change. If we wanted to break everything down to single PRs for single root causes, there's probably like 4 or more PRs in this PR already (which is why I assumed tackling another relatively small one would be reasonable, since this PR seemed very goal oriented). Is the gist that we should go put up PRs for some of these things separately and merge them first so this change doesn't break code and that'll be easier to review? Because as I've said a few times - I have a fix for the falsy type fact calculation thing in my version of this PR, it's already tested and good in all extended suites, and it's quite a small fix, very easy to have stand alone.

> Why are we even doing getNonNullableTypeIfNeeded in for-in when the statement is perfectly happy with a null or undefined right hand side value?

I honestly don't know, but comments in `in` operator checking (assuming they also apply to for-in) seem to indicate it's probably a quirk of history as the spec evolved and tradeoffs were made. Specifically, the check is written a bit backwards to try to accommodate permissiveness for generics (which is a thing because we used to never narrow generics), which is what necessitates pulling out and ignoring the nullish types. Also probably for better error messages. It doesn't really matter, since that's not fundamentally the problem, it's just one of a handful of ways to witness it (and happens to be the one our test suite caught), as my prior examples show - the last related break I've shown doesn't even have a `for` loop.

> And, even worse, why are we removing null and undefined from the right hand side in CFA when the statement doesn't actually do that. 

That actually makes total sense, imo - the loop body will never be executed if the value is nully, so if you're executing within the loop body, the value must not have been nully.

I think anything to do with for-loop logic is a wee bit off in the weeds here - our current loop checking logic works fine, and it's actually _over_-permissive when functioning as intended. And the reason it's *not* functioning as intended here is because of that type fact calculation discrepancy.

Reply:
Fine, this PR now also fixes `getFalsyFlags`.